mfkdf2/src/definitions/factor.rs

```diff
@@ -51,10 +49,26 @@ pub enum FactorType {
   OOBA(ooba::Ooba),
   Passkey(passkey::Passkey),
   Stack(stack::Stack),
+  Persisted(crate::derive::factors::persisted::Persisted),
 }
 
-impl FactorType {
-  pub fn kind(&self) -> String {
+impl FactorMetadata for FactorType {
```

Comment:
So here (or in `setup/mod.rs`), I think we can just implement `FactorSetup` for `FactorType`. I think that would get rid of the need to return `&dyn FactorSetup`, etc.

Reply:
We already implement this, it's just a matter of bypassing the `setup_mut`, and `derive_mut`.

Can we have something like

```rust
pub enum FactorType<S: FactorState> {
   // .. all the variants
}

pub trait FactorState {}

pub struct Init;
pub struct Derive;
pub struct Setup;

impl FactorState for Init {}
impl FactorState for Derive {}
impl FactorState for Setup {}
```
and now we can do something like: 
```rust
impl FactorType<Init> {
    fn derive(self) -> FactorType<Setup> { todo!() }
    fn setup(self) -> FactorType<Derive> { todo!() }
```

---

This may be overkill even. I just don't think we need to ever call the other functions that return `&dyn Trait` anymore. Those can all be inlined probably.

Reply:
started some work in https://github.com/multifactor/MFKDF2.rs/tree/refactor/typestate. core lib works perfectly, but a whole series of errors with uniffi due to uniffi not supporting generics in types.

will follow up on this later as this is not the highest priority right now.